freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/conversion/UserMetadataConverter.java

```diff
@@ -0,0 +1,124 @@
+
+@Component
+public class UserMetadataConverter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UserMetadataConverter.class);
+
+    private static final ObjectReader OBJECT_READER = new ObjectMapper().readerFor(JsonUserMetadata.class)
+            .without(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
+
+    private static final ObjectWriter OBJECT_WRITER = new ObjectMapper().writerFor(JsonUserMetadata.class);
+
+    public Optional<UserMetadata> toUserMetadata(User user) {
+        requireNonNull(user);
+
+        if (StringUtils.isBlank(user.getTitle())) {
+            // No need to warn, because this can happen under normal circumstances, if the FreeIPA user record was created before the credential
+            // update optimization entitlement was granted to the account.
+            return Optional.empty();
+        }
+
+        Optional<UserMetadata> userMetadata = Optional.empty();
+        Optional<JsonUserMetadata> jsonUserMetadata = getJsonUserMetadata(user);
+
+        if (jsonUserMetadata.isPresent() && isValidActorCrn(jsonUserMetadata.get().getCrn(), user.getUid())) {
+            Optional<UserManagementProto.WorkloadUserSyncActorMetadata> protoMetadata = decodeProtoMetadata(jsonUserMetadata.get().getMeta(), user.getUid());
+            if (protoMetadata.isPresent()) {
+                userMetadata = Optional.of(new UserMetadata(jsonUserMetadata.get().getCrn(), protoMetadata.get().getWorkloadCredentialsVersion()));
+            }
+        }
+
+        return userMetadata;
+    }
+
+    public String toUserMetadataJson(UserMetadata userMetadata) {
+        requireNonNull(userMetadata);
+        return toUserMetadataJson(userMetadata.getCrn(), userMetadata.getWorkloadCredentialsVersion());
+    }
+
+    public String toUserMetadataJson(String crn, long workloadCredentialsVersion) {
+        checkArgument(!Strings.isNullOrEmpty(crn));
+
+        // Store the metadata in a protobuf to make it easy to evolve the metadata model.
+        UserManagementProto.WorkloadUserSyncActorMetadata protoMetadata = UserManagementProto.WorkloadUserSyncActorMetadata.newBuilder()
+                .setWorkloadCredentialsVersion(workloadCredentialsVersion)
+                .build();
+        String protoMetadataBase64Encoded = BaseEncoding.base64().encode(protoMetadata.toByteArray());
+
+        // Wrap the CRN and metadata in JSON to make it easy to search for an exact match of the CRN string delimited by quotes.
+        JsonUserMetadata jsonUserMetadata = new JsonUserMetadata(crn, protoMetadataBase64Encoded);
+        try {
+            return OBJECT_WRITER.writeValueAsString(jsonUserMetadata);
+        } catch (JsonProcessingException e) {
+            throw new IllegalStateException(String.format("Unable to serialize %s as JSON: ", jsonUserMetadata), e);
+        }
+    }
+
+    private Optional<JsonUserMetadata> getJsonUserMetadata(User user) {
+        Optional<JsonUserMetadata> jsonUserMetadata = Optional.empty();
```

Comment:
please drop the variable, no need for it

Reply:
Done.